src/types.ts

```
@@ -0,0 +1,556 @@
+
+  /**
+   * Return a TokenBalanceMap from js object.
+   *
+   * @remarks
+   * Js object is expected to have the format:
+   * ```
+   * {
+   *   token1: {unlocked: n, locked: m},
+   *   token2: {unlocked: a, locked: b, lockExpires: c},
+   *   token3: {unlocked: x, locked: y, unlockedAuthorities: z, lockedAuthorities: w},
+   * }
+   * ```
+   *
+   * @param tokenBalanceMap - The js object to convert to a TokenBalanceMap
+   * @returns - The new TokenBalanceMap object
+   */
+  static fromStringMap(tokenBalanceMap: StringMap<StringMap<number | Authorities>>): TokenBalanceMap {
+    const obj = new TokenBalanceMap();
+    for (const [tokenId, balance] of Object.entries(tokenBalanceMap)) {
+      obj.set(tokenId, new Balance(
+        balance.totalSent as number,
+        balance.unlocked as number,
+        balance.locked as number,
+        // @ts-ignore
+        balance.lockExpires || null,
+        balance.unlockedAuthorities,
+        balance.lockedAuthorities,
+      ));
+    }
+    return obj;
+  }
+
+  /**
+   * Merge two TokenBalanceMap objects, merging the balances for each token.
+   *
+   * @param balanceMap1 - First TokenBalanceMap
+   * @param balanceMap2 - Second TokenBalanceMap
+   * @returns The merged TokenBalanceMap
+   */
+  static merge(balanceMap1: TokenBalanceMap, balanceMap2: TokenBalanceMap): TokenBalanceMap {
+    if (!balanceMap1) return balanceMap2.clone();
+    if (!balanceMap2) return balanceMap1.clone();
+    const mergedMap = balanceMap1.clone();
+    for (const [token, balance] of balanceMap2.iterator()) {
+      const finalBalance = Balance.merge(mergedMap.get(token), balance);
+      mergedMap.set(token, finalBalance);
+    }
+    return mergedMap;
+  }
+
+  /**
+   * Create a TokenBalanceMap from a TxOutput.
+   *
+   * @param output - The transaction output
+   * @returns The TokenBalanceMap object
+   */
+  static fromTxOutput(output: TxOutput): TokenBalanceMap {
+    // TODO check if output.decoded exists, else return null
```

Is this going to be done in this PR?

---

We can actually throw an error here, done in 4f2cecd11cae5de57147cd432a3da74150656d70